src/main/java/minijava/ir/optimize/JmpBlockRemover.java

```diff
@@ -20,14 +19,14 @@ public boolean optimize(Graph graph) {
     this.graph = graph;
     this.hasChanged = false;
     BackEdges.enable(graph);
-    graph.walkTopological(this);
+    graph.walkPostorder(this);
     BackEdges.disable(graph);
     return hasChanged;
   }
 
   @Override
   public void visit(Block block) {
-    if (!hasChanged && isJmpBlock(block) && hasOnePredecessor(block)) {
+    if (isJmpBlock(block) && hasOnePredecessor(block)) {
```

Comment:
Yep, at least I hope so. Back edges seem problematic, but I've not had any problems so far.